pages/mesosphere/dcos/2.2/storage/external-storage/csi/index.md

```diff
@@ -0,0 +1,119 @@
+---
+layout: layout.pug
+navigationTitle:  CSI Storage
+title: CSI Storage
+menuWeight: 20
+excerpt: Using CSI volumes with Marathon
+render: mustache
+model: /mesosphere/dcos/2.2/data.yml
+beta: false
+enterprise: false
+---
+
+DC/OS integrates with storage backends that use the Container Storage Interface (CSI) to interact with the cluster orchestrator (DC/OS in this case). This recently-developed standard is gaining wide adoption in the cloud-native ecosystem. If you want a storage solution which will be developed and supported long into the future, it may be wise to choose one with CSI support.
+
+Note that CSI volumes in DC/OS currently must be pre-provisioned before they are used by a service; dynamic provisioning of CSI volumes on demand is not yet supported. If you require dynamic provisioning, we currently recommend that you use [DVDI volumes][1] in DC/OS.
+ 
+# Creating an application with a CSI volume
+
+## Marathon app definition
+
+You can specify a CSI volume in your [Marathon app definition][6].
+
+Find below an example task which mounts a pre-provisioned CSI volume; it may be used as a test to verify that your CSI-based storage backend is working correctly. The `cmd` in this app definition appends the output of the `date` command to `test.txt`, reads the file, and then exits. Marathon will repeatedly re-launch the app after it finishes. You will know that the external volume is being used correctly if you see that the logs of successive runs of the application show more and more lines of `date` output; this mean that the same CSI external volume is mounted and written to each time.
+
+```json
+{
+  "id": "csi-volume-test-app",
+  "instances": 1,
+  "cpus": 0.1,
+  "mem": 32,
+  "cmd": "date >> test-csi-volume/test.txt; cat test-csi-volume/test.txt",
+  "container": {
+    "type": "MESOS",
+    "volumes": [
+      {
+        "containerPath": "test-csi-volume/",
+        "external": {
+          "provider": "csi",
+          "name": "pre-provisioned-volume-id-001",
+          "pluginName": "csi-plugin-name",
+          "options": { 
+            "accessMode": "MULTI_NODE_MULTI_WRITER",
+            "nodeStageSecret": {
+              "username": "username_secret_001",
+              "password": "password_secret_001"
+            },
+            "nodePublishSecret": {
+              "username": "username_secret_001",
+              "password": "password_secret_001"
+            }
+          }
+        }
```

Comment:
I think that it's not required because of the `accessMode` CSI field: https://github.com/mesosphere/marathon/pull/7213

@timcharper could you confirm?

Reply:
I think `accessMode` is a CSI volume specific field while `mode` is a generic field which applies to external volume, host volume and persistent volume, we need both of the two fields. And we can need to do validation between them, e.g. if `accessMode` is `xxx_READER_ONLY` but `mode` is `RW`, we should error out.